app/js/lib/intent-parser.js

```diff
@@ -55,11 +119,94 @@ export default class IntentParser {
     return string.trim();
   }
 
-  normalise(string = '') {
-    string = string.replace(/\s+/g, ' '); // Normalise whitespaces to space.
-    string = string.trim();
-    string = string.replace(pattern.punctuation, ''); // Strip punctuations.
+  normalise(string = '', locale = this.locale) {
+    // Normalise whitespaces to space.
+    return string
+      .replace(/\s+/g, ' ')
+      .trim()
+      // Strip punctuations.
+      .replace(PATTERNS[locale].punctuation, '');
+  }
+
+  /**
+   * Build the `regexps` property as an object mapping locale code to list of
+   * patterns and placeholders pairs.
+   */
+  init() {
+    Object.keys(PATTERNS).forEach((locale) => {
+      this.regexps[locale] = PATTERNS[locale].patterns.map((phrase) =>
+        this.buildRegExp(locale, phrase, PATTERNS[locale].placeholders));
+    });
+
+    console.log(this.regexps);
+  }
+
+  buildRegExp(locale = 'en', phrase = '', placeholders) {
+    phrase = this.normalise(phrase, locale);
+
+    const tokens = this.tokenise(phrase);
+    const order = {};
+    let placeholderIndex = 0;
+    let patterns = tokens.map((token) => {
+      if (token.substr(0, 1) === '[') {
+        const placeholder = token
+          .substr(1)
+          // Strip trailing `]`.
+          .replace(new RegExp('\\]$', 'u'), '');
+
+        order[placeholder] = placeholderIndex;
+        placeholderIndex++;
+
+        return placeholders[placeholder];
+      }
+
+      if (token === ' ') {
+        return '\\b \\b';
+      }
+
+      // Leading and trailing spaces are changed to word boundary.
+      return this.escape(token)
+        .replace(new RegExp('^ ', 'u'), '\\b')
+        .replace(new RegExp(' $', 'u'), '\\b');
+    });
+
+    patterns = new RegExp(`^${patterns.join('')}$`, 'iu');
+
+    return { patterns, placeholders: order };
+  }
+
+  tokenise(phrase) {
```

Comment:
Good point. I renamed it to `splitOnPlaceholders` to avoid confusion.

I previously totally forgot to use the private pattern. Thanks for the reminder. All methods and properties are now private except for `parse()` and `locale`.